enhancements/compound-bundle-constraints.md

```diff
@@ -0,0 +1,350 @@
+to remove any possible solution that includes a particular GVK, package
+at a version, or satisfies some child compound constraint from the result set.
+
+This is an example of a negation constraint of one version of a GVK,
+i.e. this GVK cannot satisfy the constraint:
+
+```yaml
+schema: olm.bundle
+name: baz.v1.0.0
+properties:
+- type: olm.none.required
+  value:
+  - type: olm.gvk.required
+    value:
+      group: foos.example.com
+      version: v1alpha1
+      kind: Foo
+```
+
+#### Nested compound constraints
+
+A nested compound constraint, one that contains at least one child compound constraint
+along with zero or more simple constraints, is evaluated from the bottom up following
+the procedures described for each above.
+
+This is an example of a disjunction of conjunctions, where one, the other, or both
+can be satisfy the constraint.
+
+```yaml
+schema: olm.bundle
+name: baz.v1.0.0
+properties:
+- type: olm.any.required
+  value:
+  - type: olm.all.required
+    value:
+    - type: olm.package.required
+      value:
+        packageName: foo
+        versionRange: '>=1.0.0'
+    - type: olm.gvk.required
+      value:
+        group: foos.example.com
+        version: v1
+        kind: Foo
+  - type: olm.all.required
+    value:
+    - type: olm.package.required
+      value:
+        packageName: foo
+        versionRange: '<1.0.0'
+    - type: olm.gvk.required
+      value:
+        group: foos.example.com
+        version: v1beta1
+        kind: Foo
+```
+
+#### Changes to OLM's resolver
+
+The [resolver package][olm-resolver-primitives] supplies the primitives "and" and "or" already,
+and uses them to combine simple GVK and package constraints into an expression.
+Compound constraints would simply extend the usage of these primitives in the nested
+case, and add the "not" primitive.
+
+### Risks and Mitigations
+
+- With the possibility for nesting comes the increased possibility that two or more nestings
+will logically conflict and result in an unsatisfiable expression.
+  - There is no quick way to determine whether an expression is unsatisfiable
+  other than by using a SAT solver, which the resolver uses internally.
+  Compound constraints must be vetted by hand.
+- Compound constraint nesting exposes an exponentiation attack vector, where a malicious
+bundle author adds deep nesting to a bundle resulting in long resolver pauses.
+  - Tools that work with bundle constraints should limit the depth of nesting.
+  I recommend the depth limit be 10 levels.
+  - Typically bundles are vetted by catalog maintainers, so deep nesting can be caught
+  by automation prior to inclusion in a catalog.
+
+## Design Details
+
+### Test Plan
+
+- operator-registry unit and e2e tests.
+- operator-lifecycle-manager unit and e2e tests.
+  - Both API and resolver unit tests.
+
+### Graduation Criteria
+
+The alpha feature gate will be removed and this feature will be on permanently by OLM v1.0.0.
```

Comment:
This decision was arbitrary. Perhaps v0.(y+2) from release is better, since we want feedback on how well this feature tackles the outlined use cases vs. improvements to the spec that result in breaking changes.